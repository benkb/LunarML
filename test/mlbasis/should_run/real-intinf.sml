infix 4 ==
val op == = Real.==
val NaN = 0.0 / 0.0;
fun sameValue (x, y) = if x == 0.0 andalso y == 0.0 then
                           Real.signBit x = Real.signBit y
                       else
                           x == y orelse (Real.isNan x andalso Real.isNan y);
datatype 'a result = VALUE of 'a
                   | OVERFLOW
                   | DOMAIN
fun resultToString (VALUE x) = LargeInt.toString x
  | resultToString OVERFLOW = "Overflow"
  | resultToString DOMAIN = "DOMAIN"
val maxFiniteAsLargeInt = IntInf.<< (0x1_ffff_ffff_ffff_f, 0w1023 - 0w52);
val () = List.app (fn (s, x, { round, floor, ceil, trunc }) =>
                      ( let val actual = VALUE (Real.toLargeInt IEEEReal.TO_NEAREST x)
                                         handle Overflow => OVERFLOW
                                              | Domain => DOMAIN
                        in if round = actual then
                               print ("toLargeInt TO_NEAREST " ^ s ^ ": OK\n")
                           else
                               print ("toLargeInt TO_NEAREST " ^ s ^ ": mismatch (expected " ^ resultToString round ^ ", but got " ^ resultToString actual ^ ")\n")
                        end
                      ; let val actual = VALUE (Real.toLargeInt IEEEReal.TO_NEGINF x)
                                         handle Overflow => OVERFLOW
                                              | Domain => DOMAIN
                        in if floor = actual then
                               print ("toLargeInt TO_NEGINF " ^ s ^ ": OK\n")
                           else
                               print ("toLargeInt TO_NEGINF " ^ s ^ ": mismatch (expected " ^ resultToString floor ^ ", but got " ^ resultToString actual ^ ")\n")
                        end
                      ; let val actual = VALUE (Real.toLargeInt IEEEReal.TO_POSINF x)
                                         handle Overflow => OVERFLOW
                                              | Domain => DOMAIN
                        in if ceil = actual then
                               print ("toLargeInt TO_POSINF " ^ s ^ ": OK\n")
                           else
                               print ("toLargeInt TO_POSINF " ^ s ^ ": mismatch (expected " ^ resultToString ceil ^ ", but got " ^ resultToString actual ^ ")\n")
                        end
                      ; let val actual = VALUE (Real.toLargeInt IEEEReal.TO_ZERO x)
                                         handle Overflow => OVERFLOW
                                              | Domain => DOMAIN
                        in if trunc = actual then
                               print ("toLargeInt TO_ZERO " ^ s ^ ": OK\n")
                           else
                               print ("toLargeInt TO_ZERO " ^ s ^ ": mismatch (expected " ^ resultToString trunc ^ ", but got " ^ resultToString actual ^ ")\n")
                        end
                      )
                  )
                  [("negInf", Real.negInf, { round = OVERFLOW, floor = OVERFLOW, ceil = OVERFLOW, trunc = OVERFLOW })
                  ,("~maxFinite", ~Real.maxFinite, { round = VALUE (~maxFiniteAsLargeInt), floor = VALUE (~maxFiniteAsLargeInt), ceil = VALUE (~maxFiniteAsLargeInt), trunc = VALUE (~maxFiniteAsLargeInt) })
                  ,("~3.5", ~3.5, { round = VALUE ~4, floor = VALUE ~4, ceil = VALUE ~3, trunc = VALUE ~3 })
                  ,("~3.25", ~3.25, { round = VALUE ~3, floor = VALUE ~4, ceil = VALUE ~3, trunc = VALUE ~3 })
                  ,("~3.0", ~3.0, { round = VALUE ~3, floor = VALUE ~3, ceil = VALUE ~3, trunc = VALUE ~3 })
                  ,("~2.75", ~2.75, { round = VALUE ~3, floor = VALUE ~3, ceil = VALUE ~2, trunc = VALUE ~2 })
                  ,("~2.5", ~2.5, { round = VALUE ~2, floor = VALUE ~3, ceil = VALUE ~2, trunc = VALUE ~2 })
                  ,("~2.25", ~2.25, { round = VALUE ~2, floor = VALUE ~3, ceil = VALUE ~2, trunc = VALUE ~2 })
                  ,("~2.0", ~2.0, { round = VALUE ~2, floor = VALUE ~2, ceil = VALUE ~2, trunc = VALUE ~2 })
                  ,("~1.75", ~1.75, { round = VALUE ~2, floor = VALUE ~2, ceil = VALUE ~1, trunc = VALUE ~1 })
                  ,("~1.5", ~1.5, { round = VALUE ~2, floor = VALUE ~2, ceil = VALUE ~1, trunc = VALUE ~1 })
                  ,("~1.25", ~1.25, { round = VALUE ~1, floor = VALUE ~2, ceil = VALUE ~1, trunc = VALUE ~1 })
                  ,("~1.0", ~1.0, { round = VALUE ~1, floor = VALUE ~1, ceil = VALUE ~1, trunc = VALUE ~1 })
                  ,("~0.75", ~0.75, { round = VALUE ~1, floor = VALUE ~1, ceil = VALUE 0, trunc = VALUE 0 })
                  ,("~0.5", ~0.5, { round = VALUE 0, floor = VALUE ~1, ceil = VALUE 0, trunc = VALUE 0 })
                  ,("~0.25", ~0.25, { round = VALUE 0, floor = VALUE ~1, ceil = VALUE 0, trunc = VALUE 0 })
                  ,("~minNormalPos", ~Real.minNormalPos, { round = VALUE 0, floor = VALUE ~1, ceil = VALUE 0, trunc = VALUE 0 })
                  ,("~minPos", ~Real.minPos, { round = VALUE 0, floor = VALUE ~1, ceil = VALUE 0, trunc = VALUE 0 })
                  ,("~0.0", ~0.0, { round = VALUE 0, floor = VALUE 0, ceil = VALUE 0, trunc = VALUE 0 })
                  ,("0.0", 0.0, { round = VALUE 0, floor = VALUE 0, ceil = VALUE 0, trunc = VALUE 0 })
                  ,("minPos", Real.minPos, { round = VALUE 0, floor = VALUE 0, ceil = VALUE 1, trunc = VALUE 0 })
                  ,("minNormalPos", Real.minNormalPos, { round = VALUE 0, floor = VALUE 0, ceil = VALUE 1, trunc = VALUE 0 })
                  ,("0.25", 0.25, { round = VALUE 0, floor = VALUE 0, ceil = VALUE 1, trunc = VALUE 0 })
                  ,("0.5", 0.5, { round = VALUE 0, floor = VALUE 0, ceil = VALUE 1, trunc = VALUE 0 })
                  ,("0.75", 0.75, { round = VALUE 1, floor = VALUE 0, ceil = VALUE 1, trunc = VALUE 0 })
                  ,("1.0", 1.0, { round = VALUE 1, floor = VALUE 1, ceil = VALUE 1, trunc = VALUE 1 })
                  ,("1.25", 1.25, { round = VALUE 1, floor = VALUE 1, ceil = VALUE 2, trunc = VALUE 1 })
                  ,("1.5", 1.5, { round = VALUE 2, floor = VALUE 1, ceil = VALUE 2, trunc = VALUE 1 })
                  ,("1.75", 1.75, { round = VALUE 2, floor = VALUE 1, ceil = VALUE 2, trunc = VALUE 1 })
                  ,("2.0", 2.0, { round = VALUE 2, floor = VALUE 2, ceil = VALUE 2, trunc = VALUE 2 })
                  ,("2.25", 2.25, { round = VALUE 2, floor = VALUE 2, ceil = VALUE 3, trunc = VALUE 2 })
                  ,("2.5", 2.5, { round = VALUE 2, floor = VALUE 2, ceil = VALUE 3, trunc = VALUE 2 })
                  ,("2.75", 2.75, { round = VALUE 3, floor = VALUE 2, ceil = VALUE 3, trunc = VALUE 2 })
                  ,("3.0", 3.0, { round = VALUE 3, floor = VALUE 3, ceil = VALUE 3, trunc = VALUE 3 })
                  ,("3.25", 3.25, { round = VALUE 3, floor = VALUE 3, ceil = VALUE 4, trunc = VALUE 3 })
                  ,("3.5", 3.5, { round = VALUE 4, floor = VALUE 3, ceil = VALUE 4, trunc = VALUE 3 })
                  ,("maxFinite", Real.maxFinite, { round = VALUE maxFiniteAsLargeInt, floor = VALUE maxFiniteAsLargeInt, ceil = VALUE maxFiniteAsLargeInt, trunc = VALUE maxFiniteAsLargeInt })
                  ,("posInf", Real.posInf, { round = OVERFLOW, floor = OVERFLOW, ceil = OVERFLOW, trunc = OVERFLOW })
                  ,("NaN", NaN, { round = DOMAIN, floor = DOMAIN, ceil = DOMAIN, trunc = DOMAIN })
                  ];
