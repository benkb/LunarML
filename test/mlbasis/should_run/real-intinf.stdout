toLargeInt TO_NEAREST negInf: OK
toLargeInt TO_NEGINF negInf: OK
toLargeInt TO_POSINF negInf: OK
toLargeInt TO_ZERO negInf: OK
toLargeInt TO_NEAREST ~maxFinite: OK
toLargeInt TO_NEGINF ~maxFinite: OK
toLargeInt TO_POSINF ~maxFinite: OK
toLargeInt TO_ZERO ~maxFinite: OK
toLargeInt TO_NEAREST ~3.5: OK
toLargeInt TO_NEGINF ~3.5: OK
toLargeInt TO_POSINF ~3.5: OK
toLargeInt TO_ZERO ~3.5: OK
toLargeInt TO_NEAREST ~3.25: OK
toLargeInt TO_NEGINF ~3.25: OK
toLargeInt TO_POSINF ~3.25: OK
toLargeInt TO_ZERO ~3.25: OK
toLargeInt TO_NEAREST ~3.0: OK
toLargeInt TO_NEGINF ~3.0: OK
toLargeInt TO_POSINF ~3.0: OK
toLargeInt TO_ZERO ~3.0: OK
toLargeInt TO_NEAREST ~2.75: OK
toLargeInt TO_NEGINF ~2.75: OK
toLargeInt TO_POSINF ~2.75: OK
toLargeInt TO_ZERO ~2.75: OK
toLargeInt TO_NEAREST ~2.5: OK
toLargeInt TO_NEGINF ~2.5: OK
toLargeInt TO_POSINF ~2.5: OK
toLargeInt TO_ZERO ~2.5: OK
toLargeInt TO_NEAREST ~2.25: OK
toLargeInt TO_NEGINF ~2.25: OK
toLargeInt TO_POSINF ~2.25: OK
toLargeInt TO_ZERO ~2.25: OK
toLargeInt TO_NEAREST ~2.0: OK
toLargeInt TO_NEGINF ~2.0: OK
toLargeInt TO_POSINF ~2.0: OK
toLargeInt TO_ZERO ~2.0: OK
toLargeInt TO_NEAREST ~1.75: OK
toLargeInt TO_NEGINF ~1.75: OK
toLargeInt TO_POSINF ~1.75: OK
toLargeInt TO_ZERO ~1.75: OK
toLargeInt TO_NEAREST ~1.5: OK
toLargeInt TO_NEGINF ~1.5: OK
toLargeInt TO_POSINF ~1.5: OK
toLargeInt TO_ZERO ~1.5: OK
toLargeInt TO_NEAREST ~1.25: OK
toLargeInt TO_NEGINF ~1.25: OK
toLargeInt TO_POSINF ~1.25: OK
toLargeInt TO_ZERO ~1.25: OK
toLargeInt TO_NEAREST ~1.0: OK
toLargeInt TO_NEGINF ~1.0: OK
toLargeInt TO_POSINF ~1.0: OK
toLargeInt TO_ZERO ~1.0: OK
toLargeInt TO_NEAREST ~0.75: OK
toLargeInt TO_NEGINF ~0.75: OK
toLargeInt TO_POSINF ~0.75: OK
toLargeInt TO_ZERO ~0.75: OK
toLargeInt TO_NEAREST ~0.5: OK
toLargeInt TO_NEGINF ~0.5: OK
toLargeInt TO_POSINF ~0.5: OK
toLargeInt TO_ZERO ~0.5: OK
toLargeInt TO_NEAREST ~0.25: OK
toLargeInt TO_NEGINF ~0.25: OK
toLargeInt TO_POSINF ~0.25: OK
toLargeInt TO_ZERO ~0.25: OK
toLargeInt TO_NEAREST ~minNormalPos: OK
toLargeInt TO_NEGINF ~minNormalPos: OK
toLargeInt TO_POSINF ~minNormalPos: OK
toLargeInt TO_ZERO ~minNormalPos: OK
toLargeInt TO_NEAREST ~minPos: OK
toLargeInt TO_NEGINF ~minPos: OK
toLargeInt TO_POSINF ~minPos: OK
toLargeInt TO_ZERO ~minPos: OK
toLargeInt TO_NEAREST ~0.0: OK
toLargeInt TO_NEGINF ~0.0: OK
toLargeInt TO_POSINF ~0.0: OK
toLargeInt TO_ZERO ~0.0: OK
toLargeInt TO_NEAREST 0.0: OK
toLargeInt TO_NEGINF 0.0: OK
toLargeInt TO_POSINF 0.0: OK
toLargeInt TO_ZERO 0.0: OK
toLargeInt TO_NEAREST minPos: OK
toLargeInt TO_NEGINF minPos: OK
toLargeInt TO_POSINF minPos: OK
toLargeInt TO_ZERO minPos: OK
toLargeInt TO_NEAREST minNormalPos: OK
toLargeInt TO_NEGINF minNormalPos: OK
toLargeInt TO_POSINF minNormalPos: OK
toLargeInt TO_ZERO minNormalPos: OK
toLargeInt TO_NEAREST 0.25: OK
toLargeInt TO_NEGINF 0.25: OK
toLargeInt TO_POSINF 0.25: OK
toLargeInt TO_ZERO 0.25: OK
toLargeInt TO_NEAREST 0.5: OK
toLargeInt TO_NEGINF 0.5: OK
toLargeInt TO_POSINF 0.5: OK
toLargeInt TO_ZERO 0.5: OK
toLargeInt TO_NEAREST 0.75: OK
toLargeInt TO_NEGINF 0.75: OK
toLargeInt TO_POSINF 0.75: OK
toLargeInt TO_ZERO 0.75: OK
toLargeInt TO_NEAREST 1.0: OK
toLargeInt TO_NEGINF 1.0: OK
toLargeInt TO_POSINF 1.0: OK
toLargeInt TO_ZERO 1.0: OK
toLargeInt TO_NEAREST 1.25: OK
toLargeInt TO_NEGINF 1.25: OK
toLargeInt TO_POSINF 1.25: OK
toLargeInt TO_ZERO 1.25: OK
toLargeInt TO_NEAREST 1.5: OK
toLargeInt TO_NEGINF 1.5: OK
toLargeInt TO_POSINF 1.5: OK
toLargeInt TO_ZERO 1.5: OK
toLargeInt TO_NEAREST 1.75: OK
toLargeInt TO_NEGINF 1.75: OK
toLargeInt TO_POSINF 1.75: OK
toLargeInt TO_ZERO 1.75: OK
toLargeInt TO_NEAREST 2.0: OK
toLargeInt TO_NEGINF 2.0: OK
toLargeInt TO_POSINF 2.0: OK
toLargeInt TO_ZERO 2.0: OK
toLargeInt TO_NEAREST 2.25: OK
toLargeInt TO_NEGINF 2.25: OK
toLargeInt TO_POSINF 2.25: OK
toLargeInt TO_ZERO 2.25: OK
toLargeInt TO_NEAREST 2.5: OK
toLargeInt TO_NEGINF 2.5: OK
toLargeInt TO_POSINF 2.5: OK
toLargeInt TO_ZERO 2.5: OK
toLargeInt TO_NEAREST 2.75: OK
toLargeInt TO_NEGINF 2.75: OK
toLargeInt TO_POSINF 2.75: OK
toLargeInt TO_ZERO 2.75: OK
toLargeInt TO_NEAREST 3.0: OK
toLargeInt TO_NEGINF 3.0: OK
toLargeInt TO_POSINF 3.0: OK
toLargeInt TO_ZERO 3.0: OK
toLargeInt TO_NEAREST 3.25: OK
toLargeInt TO_NEGINF 3.25: OK
toLargeInt TO_POSINF 3.25: OK
toLargeInt TO_ZERO 3.25: OK
toLargeInt TO_NEAREST 3.5: OK
toLargeInt TO_NEGINF 3.5: OK
toLargeInt TO_POSINF 3.5: OK
toLargeInt TO_ZERO 3.5: OK
toLargeInt TO_NEAREST maxFinite: OK
toLargeInt TO_NEGINF maxFinite: OK
toLargeInt TO_POSINF maxFinite: OK
toLargeInt TO_ZERO maxFinite: OK
toLargeInt TO_NEAREST posInf: OK
toLargeInt TO_NEGINF posInf: OK
toLargeInt TO_POSINF posInf: OK
toLargeInt TO_ZERO posInf: OK
toLargeInt TO_NEAREST NaN: OK
toLargeInt TO_NEGINF NaN: OK
toLargeInt TO_POSINF NaN: OK
toLargeInt TO_ZERO NaN: OK
